### ISSUE TITLE
BIGDL-LM Acceleration for chatglm3-6b

### ISSUE THREAD

**Issue by HuskyLYL** [intel-analytics/BigDL#10309]
I’ve been using BIGDL-LM to accelerate the chatglm3-6b model. However, I’m curious about the speed. Is the current speed considered normal?
Here are the hardware details:
+ Graphics Card: Intel Corporation TigerLake-LP GT2
+ Processor: 11th Gen Intel® Core™ i5-1135G7 @ 2.40GHz
 Architecture: x86_64
 CPU Cores: 8
 Threads: 16

```python
import time

from bigdl.llm.transformers import AutoModel

from transformers import AutoTokenizer

import sys
     
model_path = "/home/ty/chatglm3-demo/chatglm3-6b"

model = AutoModel.from_pretrained(model_path,load_in_4bit=True,trust_remote_code=True)

tokenizer = AutoTokenizer.from_pretrained(model_path,trust_remote_code=True)

text = "If you are a tour guide and a tourist asks you, 'Do you know Yuelu Academy?'What should you do, how should you answer?"

input_ids = tokenizer.encode(text, return_tensors="pt")

st = time.time()

response, history = model.chat(tokenizer, text, history=[])

end = time.time()

print(f'Inference time: {end-st} s')

print('-'*20, 'Prompt', '-'*20)

print(text)

print('-'*20, 'Output', '-'*20)

print(response)

sys.exit()


```

![8569bad769e59fe8835a530fd9c3e86d](https://github.com/intel-analytics/BigDL/assets/91365773/ac56551f-67e5-4253-af78-a7d04050c420)

**Comment by HuskyLYL** [comment]
My latency varies between 20 and 40 seconds, and it’s a simple query without any historical conversation messages. Is there a way to further optimize my model’s execution speed without altering the hardware conditions?

**Comment by Romanticoseu** [comment]
Based on the `model_path = "/home/ty/chatglm3-demo/chatglm3-6b" `in your code, could you confirm if you're running the script on WSL? Also, could you provide details about the memory of your machine and the memory allocated to WSL? 
Additionally, you can try using `with torch.inference_model` to see if there's any improvement.
Example code:
```python
with torch.inference_mode():
        prompt = CHATGLM_V3_PROMPT_FORMAT.format(prompt=args.prompt)
        input_ids = tokenizer.encode(prompt, return_tensors="pt")
        st = time.time()
        # if your selected model is capable of utilizing previous key/value attentions
        # to enhance decoding speed, but has `"use_cache": false` in its model config,
        # it is important to set `use_cache=True` explicitly in the `generate` function
        # to obtain optimal performance with BigDL-LLM INT4 optimizations
        output = model.generate(input_ids,
                                max_new_tokens=args.n_predict)
        end = time.time()
        output_str = tokenizer.decode(output[0], skip_special_tokens=True)
        print(f'Inference time: {end-st} s')
        print('-'*20, 'Prompt', '-'*20)
        print(prompt)
        print('-'*20, 'Output', '-'*20)
        print(output_str)
```

**Comment by HuskyLYL** [comment]
> Based on the `model_path = "/home/ty/chatglm3-demo/chatglm3-6b" `in your code, could you confirm if you're running the script on WSL? Also, could you provide details about the memory of your machine and the memory allocated to WSL? Additionally, you can try using `with torch.inference_model` to see if there's any improvement. Example code:
> 
> ```python
> with torch.inference_mode():
>         prompt = CHATGLM_V3_PROMPT_FORMAT.format(prompt=args.prompt)
>         input_ids = tokenizer.encode(prompt, return_tensors="pt")
>         st = time.time()
>         # if your selected model is capable of utilizing previous key/value attentions
>         # to enhance decoding speed, but has `"use_cache": false` in its model config,
>         # it is important to set `use_cache=True` explicitly in the `generate` function
>         # to obtain optimal performance with BigDL-LLM INT4 optimizations
>         output = model.generate(input_ids,
>                                 max_new_tokens=args.n_predict)
>         end = time.time()
>         output_str = tokenizer.decode(output[0], skip_special_tokens=True)
>         print(f'Inference time: {end-st} s')
>         print('-'*20, 'Prompt', '-'*20)
>         print(prompt)
>         print('-'*20, 'Output', '-'*20)
>         print(output_str)
> ```

It's on my aipc host, I don't use WSL. I've tried your method, and the results are barely noticeable

**Comment by Romanticoseu** [comment]
You can try warmup and then check if the inference speed improves.
Certainly, the `max_new_tokens` parameter can lead to varying inference times; instead, you can utilize `max_length` to regulate it.
```python
import time

from bigdl.llm.transformers import AutoModel

from transformers import AutoTokenizer

import sys
import torch

warm_up = 1
num_trials = 3

model_path = r"C:\Users\ml_so\llm\models\chatglm3-6b"
model = AutoModel.from_pretrained(model_path,load_in_4bit=True,trust_remote_code=True)

tokenizer = AutoTokenizer.from_pretrained(model_path,trust_remote_code=True)

text = "If you are a tour guide and a tourist asks you, 'Do you know Yuelu Academy?'What should you do, how should you answer?"


# input_ids = tokenizer.encode(text, return_tensors="pt")
for i in range(num_trials + warm_up):
    st = time.time()
    response, history = model.chat(tokenizer, text, history=[], max_length=128)
    end = time.time()
    if i < warm_up:
        print(f"warm_up time: {end-st} s")
    else:
        print(f'Inference time: {end-st} s')
    # print('-'*20, 'Prompt', '-'*20)
    # print(text)
    # print('-'*20, 'Output', '-'*20)
    print(response)
```
